### PR TITLE
Switch to ruff instead of flake8.

### PR DESCRIPTION
**Pull request by clokep**
See https://github.com/charliermarsh/ruff/

> An extremely fast Python linter, written in Rust.

With this PR I mostly aimed at reproducing our current `flake8` config in `ruff` -- there are a lot of other checks to enable, but I see that as follow-up work. The goal is to improve developer productivity via:

* Being much faster.
* Having a better `--fix` flag that just works for most things.

Note that ruff can [also replace `isort`](https://github.com/charliermarsh/ruff/#isort-i), but it doesn't implement [custom sections](https://pycqa.github.io/isort/docs/configuration/custom_sections_and_ordering.html).

----

This shaves about 30s off CI jobs (`ruff` runs in 1s vs. `flake8` running in ~30s). On my system `ruff` runs in ~2s, while `flake8` takes somewhere between 60 - 90s.

### CODE REVIEW

**Comment by clokep** [comment]
Meh, looks like this pulls an action in from somewhere else to do linting -- https://github.com/matrix-org/backend-meta/blob/1a8479c9088301133b58cc77421d5e8df0e0214e/.github/workflows/python-poetry-ci.yml#L36

Will need to figure out if changing that will break everything.

**Comment by clokep** [comment]
I ran `flake8` on `develop`:

```
$ time flake8 synapse docker tests scripts-dev contrib synmark stubs .ci
flake8 synapse docker tests scripts-dev contrib synmark stubs .ci  62.45s user 1.17s system 95% cpu 1:06.33 total
```

And then `ruff` on this branch:
```
$time ruff synapse docker tests scripts-dev contrib synmark stubs .ci  
ruff synapse docker tests scripts-dev contrib synmark stubs .ci  1.36s user 0.32s system 484% cpu 0.347 total
```

That seems worth pursuing...

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -53,7 +53,7 @@ jobs:
       - run: scripts-dev/check_schema_delta.py --force-colors
 
   lint:
-    uses: "matrix-org/backend-meta/.github/workflows/python-poetry-ci.yml@v1"
+    uses: "matrix-org/backend-meta/.github/workflows/python-poetry-ci.yml@1ee1eb9eb3e418f9631e52e8a7ab371ca26f163d"
```
Comment (clokep) [inline .github/workflows/tests.yml:56]: We need to release after matrix-org/backend-meta#15 is merged?

File: stubs/frozendict.pyi [hunk 2]
```diff
@@ -14,6 +14,8 @@
 
 # Stub for frozendict.
 
+from __future__ import annotations
```
Comment (clokep) [inline stubs/frozendict.pyi:17]: We do seem to actually need these, I'm not sure why `flake8` was running OK without them. Maybe it wasn't even checking `*.pyi` files?

**Review by reivilibre — APPROVED** [review]
Excellent.

Would love to replace isort as well. I'm personally not fussed about custom sections but I am fussed about it taking forever for a no-op change. (I even see that isort seems to create a temporary file for every file it processes!)

**Comment by clokep** [comment]
> Would love to replace isort as well. I'm personally not fussed about custom sections but I am fussed about it taking forever for a no-op change. (I even see that isort seems to create a temporary file for every file it processes!)

I agree, I think we can do this in the future though?

**Comment by reivilibre** [comment]
yeah, definitely, that can be done later :)

File: .github/workflows/tests.yml (hunk 1 above)
Comment (clokep) [reply .github/workflows/tests.yml:56]:
```suggestion
    uses: "matrix-org/backend-meta/.github/workflows/python-poetry-ci.yml@v2"
```

**Comment by clokep** [comment]
I also updated the docs: 14abf22dd696129f69cd1a14b2044a4ca6cb1972

**Comment by charliermarsh** [comment]
This is really cool to see. I hope commenting here doesn't come off as intrusive, but just wanted to say that if you run into any issues or if there are any changes or improvements to Ruff that would be particularly high-impact for the project, I'm very receptive to feedback over on the repo and try to turn that kind of stuff around pretty quickly.

**Comment by clokep** [comment]
> This is really cool to see. I hope commenting here doesn't come off as intrusive, but just wanted to say that if you run into any issues or if there are any changes or improvements to Ruff that would be particularly high-impact for the project, I'm very receptive to feedback over on the repo and try to turn that kind of stuff around pretty quickly.

Thanks for the offer! I haven't noticed anything major, but we'll definitely file issues if we see problems!

**Comment by clokep** [comment]
This had some follow-ups: #14741, and #14821. Hopefully we're not in the same state as with `flake8`.

**Comment by charliermarsh** [comment]
I went ahead and disabled the update-check by-default. I've received enough feedback that it clearly shouldn't be on by-default. Sorry for the noise!

**Pull request merged**